test/functional/mempool_updatefromblock.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test mempool descendants/ancestors information update.
+
+Test mempool update of transaction descendants/ancestors information (count, size)
+when transactions have been re-added from a disconnected block to the mempool.
+"""
+import time
+
+from decimal import Decimal
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+
+class MempoolUpdateFromBlockTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [['-limitdescendantsize=1000', '-limitancestorsize=1000']]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', end_address='', fee=Decimal(0.00100000)):
+        """Create a tournament (a type of directed graph) of transactions and use it for testing.
+
+        Keyword arguments:
+        size -- the order N of the tournament which is equal to the number of the created transactions
+        n_tx_to_mine -- the number of transaction that should be mined into a block
+
+        If all of the N created transactions tx[0]..tx[N-1] reside in the mempool,
+        the following holds:
+            the tx[K] transaction:
+            - has N-K descendants (including this one), and
+            - has K+1 ancestors (including this one)
+
+        More details: https://en.wikipedia.org/wiki/Tournament_(graph_theory)
+        """
+
+        if not start_input_txid:
+            start_input_txid = self.nodes[0].getblock(self.nodes[0].getblockhash(1))['tx'][0]
+
+        if not end_address:
+            end_address = self.nodes[0].getnewaddress()
+
+        first_block_hash = ''
+        tx_id = []
+        tx_size = []
+        self.log.info('Creating {} transactions...'.format(size))
+        for i in range(0, size):
+            self.log.debug('Preparing transaction #{}...'.format(i))
+            # Prepare inputs.
+            if i == 0:
+                inputs = [{'txid': start_input_txid, 'vout': 0}]
+                inputs_value = self.nodes[0].gettxout(start_input_txid, 0)['value']
+            else:
+                inputs = []
+                inputs_value = 0
+                for j, tx in enumerate(tx_id[0:i]):
+                    vout = i - j - 1
```

Comment:
"Transaction N is children of every previous transactions at their output N-1"

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/18485#issuecomment-619077646).